### ISSUE TITLE
RE-INVITE sent before ICE gathering is complete

### ISSUE THREAD

**Issue by shazzzzzzzz** [versatica/JsSIP#499]
**scenario 1-** 

- A calls B. 
     - Audio only. 
     - A supports Bundle and Trickle ICE. 
     - No Bundle or Trickle ICE support on B.
- A turns on camera. 
     - Application internally adds a video track to the existing stream on A and starts re-negotiation.

**Expected behavior** - Since JsSIP does not support trickle ICE, it must trigger a RE-INVITE once it gathers all the ICE candidate for the new video track.
**Observed behavior** - RE-INVITE is sent out before new ICE candidates could be gathered. B is unaware of A's candidates, hence no RTP streaming possible.

From my code analysis(if it helps) - When A triggers re-negotiation, iceGatheringState is already "complete" upon setLocalDescription invocation. Due to this create offer's success callback is invoked before ICE gathering is actually completed. 

This could impact another scenario as well as below
**Scenario 2-**

- B calls A. Audio 'sendrecv' and video 'inactive'
- B turns on camera. Replaces existing stream with a new stream containing video tracks and triggers re-negotiation with new ICE candidates
- A responds back with 200 OK, however I still see ICE gathering being completed after 200 OK is sent out.

**Expected behavior** - 
200 OK must be sent after ICE gathering is complete
                                       OR 
Should 'A' even gather ICE candidates? since it could use the same candidates from initial call setup. With scenario 2(B calls A. Audio 'sendrecv' and video 'inactive') and A turning on camera, I still see new ICE candidates being gathered for RE-INVITE to be sent out. I am not sure on the expected behavior here. 

Attached are the browser logs for both scenarios on A. 

[scenario_1.txt](https://github.com/versatica/JsSIP/files/1682353/scenario_1.txt)
[scenario_2.txt](https://github.com/versatica/JsSIP/files/1682354/scenario_2.txt)

**Comment by jmillan** [comment]
Hi,

May you register a callback to the `RTCSession::connection` ` negotiationneeded` event once you add the video track and actually renegotiate from such callback? Looks like the peerconnection is not totally ready yet when re-inviting. Ie:


```javascript
function renegotiate()
{
  session.renegotiate();
  session.connection.removeEventListener('negotiationneeded', renegotiate);
}

// Add video track here...

session.connection.addEventListener('negotiationneeded', renegotiate);
```

Thanks

**Comment by shazzzzzzzz** [comment]
Hi,

I tried the solution you suggested and here are the results.

**Scenario 1 -**
Technically works with the changes you proposed. However,  I see ICE candidate gathering debug after RE-INVITE is sent out and surprisingly the RE-INVITE message has the gathered candidates for video SDP. Please refer the new logs. I am still not clear if the event '**icecandidate**' is triggered after RE-INVITE is sent out??(based on logs)

**Scenario 2 -** 
A calls B. Audio 'sendrecv' and video 'inactive'
A turns on camera. Replaces existing stream with a new stream containing video tracks and triggers re-negotiation with same ICE candidates as before(initial call setup)
B responds back with 200 OK, with same ICE candidates as before(initial call setup)

works as expected(when the caller initiates RE-INVITE)

Consider below scenario- 

**Scenario 3** 

- B calls A. Audio 'sendrecv' and video 'inactive'
- First re-negotiation(callee initiates RE-INVTE) - 
    - A turns on camera. Replaces existing stream with a new stream containing video tracks and triggers re-negotiation. However I still see ICE gathering being completed after RE-INVITE is sent out. The new ICE candidate sends the video stream to B, not the one present in 200 OK. 200 OK has the ICE candidate from initial call setup!
    - B responds back with 200 OK with same ICE candidates as before(initial call setup)
- Second re-negotiation - 
    - B turns on camera. Replaces existing stream with a new stream containing video tracks and triggers re-negotiation with same ICE candidates as before(initial call setup/first negotiation)
    - A responds back with 200 OK with 2 candidates for video with same priority. First candidate from initial call setup and second one gathered during first re-negotiation.

**Expected behavior** - During first re-negotiation, A should not gather ICE candidates for video SDP since it could reuse the existing one from initial call setup. This issue leads to the strange behavior
 in second re-negotiation where in A sends the 200 OK with two candidates of same priority. 

To summarize, when the callee triggers RE-INVITE, no ICE gathering must happen if there is a candidate available from initial call setup. When the caller triggers RE-INVTE, ICE candidates are re-used from initial call setup as expected
	
Attached are the logs
[scenario_1_working.txt](https://github.com/versatica/JsSIP/files/1686059/scenario_1_working.txt)
[scenario_3.txt](https://github.com/versatica/JsSIP/files/1686060/scenario_3.txt)

**Comment by jmillan** [comment]
Hi,

> Expected behavior - During first re-negotiation, A should not gather ICE candidates for video SDP since it could reuse the existing one from initial call setup. This issue leads to the strange behavior
in second re-negotiation where in A sends the 200 OK with two candidates of same priority.

Whether it should gather new candidates or not is up to the [specification](https://www.w3.org/TR/webrtc/). On the other hand JsSIP is not responsible of such implementation but the browser on top of which is running.

You may want to use bare RTCPeerconnection instances, without any signalling at all in order to probe that the implementation does not satisfy the specification and report to the corresponding browser vendor.

Apart from that, we should notice somewhere the fact that if a stream is modified or another action is done over the RTCPeerconnection instance which will trigger a `renegotiationneeded`, such process should be respected.

**Comment by shazzzzzzzz** [comment]
Hi,

Thanks for clarification on 're-use of ICE candidates'. I will investigate further using bare RTCPeerConnection and contact browser vendor for any clarifications.

Coming to the current issue, even with the code changes you suggested, the case of missing ICE candidates could be reproduced through a simple audio call as below

- A calls B - audio only
- A triggers re-invite with new audio stream(**iceRestart = true**)
- A sends out RE-INVTE before ICE gathering is complete
- B responds back with 200 OK before ICE gathering is complete 
- Call fails with RTP timeout

**Expected behavior** - Both A and B must have ICE candidates in the RE-INVTE and 200 OK messages respectively
**Observed behavior** - Even with the code changes, ICE candidates are missing which results in RTP timeout.

If it helps, I still suspect the root cause to be the 'iceGatheringState' being 'complete' when the renegotaition process starts. The 'iceGatheringState' is never reset at the start of re-negotiation process. Success callback of 'setLocalDescription' is invoked even before ICE gathering is actually complete.  

Attached are the browser logs on A and B

[browser_logs_A.txt](https://github.com/versatica/JsSIP/files/1695771/browser_logs_A.txt)
[browser_logs_B.txt](https://github.com/versatica/JsSIP/files/1695772/browser_logs_B.txt)

**Comment by jmillan** [comment]
Closing because the issue is totally related to WebRTC implementation, not JsSIP.
